SwiftDriver: add support for baremetal targets (to reflect https://github.com/apple/swift/pull/35970)

kubamracek opened the pull request.

compnerd commented on Sources/SwiftDriver/Driver/Driver.swift:
```diff
@@ -2892,12 +2892,16 @@ extension Driver {
+    guard let os else {
+      diagnosticsEngine.emit(.error_unknown_target(triple))
+      throw Driver.ErrorDiagnostics.emitted
+    }
```
Why the guard over the default case?

rauhul replied:
I'm guessing it's to avoid future reader ambiguity between Optional.none and OS.none.

compnerd replied:
That would be helpful as a comment if that is the case.  Doubly so since the handling is identical.

compnerd commented on Sources/SwiftDriver/Utilities/Triple+Platforms.swift:
```diff
@@ -282,9 +282,10 @@ extension Triple {
-    case nil:
+    guard let os else {
       fatalError("unknown OS")
+    }
```
Why not sink the `fatalError` into the `default` case?

MaxDesiatov replied:
I'd assume if `default` case is added to this `switch`, that would silence future compiler errors in this `switch` when a new case is added to `Triple.OS`.

compnerd replied:
Sure, but you can handle that with `.some` and `.none` :)

kubamracek commented:
So apparently the "none" case was at least confusing but I'd say misleading. Changed it to "noneOS" to avoid the clash with Optional.none. Added tests.

kubamracek commented:
@swift-ci please test

compnerd reviewed (approved):
Thank you for porting the work from the c++ driver!  I know I'd been slow to finish up the work :(

MaxDesiatov commented:
Is it worth retrying the Windows job? Couldn't find the exact failure reason for it.

compnerd commented:
Seems to have timed out.  I think that the machine sometimes is under higher load.

compnerd commented:
@swift-ci please test windows platform
